amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ConsumerModeUpdateIntegrationTest.java

```diff
@@ -66,218 +85,352 @@ public static Collection<Object[]> data() {
 
             } finally {
                 // Clean up
-                producer.stop(true);
-                consumerA.stop(true);
-                consumerBPrev.stop(true);
-                consumerBNext.stop(true);
+                helper.cleanup();
             }
         }
+    }
 
-        private void verifyLeasesType(DynamoDbClient ddbClient, String tableName) throws Exception {
-            ScanRequest scanRequest = ScanRequest.builder().tableName(tableName).build();
-            ScanResponse scanResponse = ddbClient.scan(scanRequest);
-            for (Map<String, AttributeValue> item : scanResponse.items()){
-                String leaseOwner = item.get("leaseOwner").s();
-                if (leaseOwner.endsWith("A")) {
-                    assertEquals(true, verifyLeaseType(item, prevMode));
-                } else if (leaseOwner.endsWith("BNext")) {
-                    assertEquals(true, verifyLeaseType(item, nextMode));
-                } else {
-                    throw new Exception("A stale lease owned by " + leaseOwner + " found after running the scenario.");
-                }
+    public class FailurePathNonParameterizedIntegrationTest {
+        @Test
+        public void consumerModeUpdateFromSingleStreamToSingleStreamUpgradeTest() throws Exception {
+            KCLMode prevMode = KCLMode.SingleStream;
+            KCLMode nextMode = KCLMode.SingleStreamUpgrade;
+            ConsumerModeUpdateIntegrationTestHelper helper = new ConsumerModeUpdateIntegrationTestHelper(prevMode,
+                    nextMode);
+            try {
+                helper.runProducerAndConsumersThenWait(Lists.newArrayList(helper.consumerA, helper.consumerBPrev));
+                helper.stopProducerAndConsumers(Lists.newArrayList(helper.consumerBPrev));
+                assertTrue(helper.validateRecordsDelivery());
+
+                helper.runProducerAndConsumersThenWait(Lists.newArrayList(helper.consumerA, helper.consumerBNext));
+                helper.stopProducerAndConsumers(Lists.newArrayList(helper.consumerBNext, helper.consumerA));
+                assertTrue(helper.validateRecordsDelivery());
+                verifyLeasesType(helper.prevKCLAppConfig.buildAsyncDynamoDbClient(),
+                        KCLAppConfig.INTEGRATION_TEST_RESOURCE_PREFIX + helper.testName,
+                        helper.consumerA.getWorkerIdentifier(), prevMode, helper.consumerBNext.getWorkerIdentifier(),
+                        nextMode);
+
+                helper.runProducerAndConsumersThenWait(Lists.newArrayList(helper.consumerA));
+                helper.stopProducerAndConsumers(Lists.newArrayList(helper.consumerA));
+                assertFalse(helper.validateRecordsDelivery());
+                assertTrue(getLeaseCountOwnedBy(helper.prevKCLAppConfig.buildAsyncDynamoDbClient(),
+                        KCLAppConfig.INTEGRATION_TEST_RESOURCE_PREFIX + helper.testName,
+                        helper.consumerBNext.getWorkerIdentifier()) > 0);
```

Comment:
What are we validating here?

Reply:
Consumer A won't steal the leases from consumer B as lease format is incompatible. Therefore verifying consumer B is keep owning >0 lease.